packages/jupyter-ai/jupyter_ai/actors/autonotebook.py

```diff
@@ -0,0 +1,235 @@
+import json
+import os
+import time
+from uuid import uuid4
+
+import ray
+from ray.util.queue import Queue
+
+from langchain.llms import BaseLLM
+from langchain.chat_models import ChatOpenAI
+from langchain.prompts import PromptTemplate
+from langchain.llms import BaseLLM
+from langchain.chains import LLMChain
+
+import nbformat
+
+from jupyter_ai.models import AgentChatMessage, HumanChatMessage
+from jupyter_ai.actors.base import BaseActor, Logger
+from jupyter_ai_magics.providers import ChatOpenAINewProvider
+
+schema = """{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "properties": {
+    "description": {
+      "type": "string"
+    },
+    "sections": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "title": {
+            "type": "string"
+          },
+          "content": {
+            "type": "string"
+          }
+        },
+        "required": ["title", "content"]
+      }
+    }
+  },
+  "required": ["sections"]
+}"""
+
+class NotebookOutlineChain(LLMChain):
+    """Chain to generates tasks."""
+
+    @classmethod
+    def from_llm(cls, llm: BaseLLM, verbose: bool=False) -> LLMChain:
+        task_creation_template = (
+            "You are an AI that creates a detailed content outline for a Jupyter notebook on a given topic.\n"
+            "Generate the outline as JSON data that will validate against this JSON schema:\n"
+            "{schema}\n"
+            "Here is a description of the notebook you will create an outline for: {description}\n"
+            "Don't include an introduction or conclusion section in the outline, focus only on sections that will need code."
+        )
+        prompt = PromptTemplate(
+            template=task_creation_template,
+            input_variables=[
+                "description",
+                "schema"
+            ],
+        )
+        return cls(prompt=prompt, llm=llm, verbose=verbose)
+
+def generate_outline(description, llm=None, verbose=False):
+    if llm is None:
+        llm = ChatOpenAINewProvider(model_id='gpt-3.5-turbo')
+    chain = NotebookOutlineChain.from_llm(llm=llm, verbose=verbose)
+    outline = chain.predict(description=description, schema=schema)
+    return json.loads(outline)
+
+class CodeImproverChain(LLMChain):
+    """Chain to generates tasks."""
+
+    @classmethod
+    def from_llm(cls, llm: BaseLLM, verbose: bool=False) -> LLMChain:
+        task_creation_template = (
+            "Improve the following code and make sure it is valid. Make sure to return the improved code only - don't give an explanation of the improvements.\n"
```

Comment:
"Improve" is ambiguous; should we specify what improvements we want?

Reply:
I played with the prompt a bit and the LLMs seem to like this phrasing. We can continue to play with this over time.